add lexer for Swan language

ansbmartin opened the pull request:
**This pull request is created to get the technical and legal approvals before submitting a PR to https://github.com/pygments/pygments**

---

This lexer supports the Swan programming language, used in Ansys Scade One software.

Swan language is defined in https://github.com/ansys/DevRelPublic/blob/main/Downloads/ScadeOne/SwanPrimer_Rev_2_1.pdf

ansbmartin committed:
add lexer for Swan language

This lexer supports the Swan programming language, used
in Ansys Scade One software.

Swan language is defined in
https://github.com/ansys/DevRelPublic/blob/main/Downloads/ScadeOne/SwanPrimer_Rev_2_1.pdf

xfornari commented on pygments/lexers/_mapping.py:
```diff
@@ -504,6 +504,7 @@
     'StanLexer': ('pygments.lexers.modeling', 'Stan', ('stan',), ('*.stan',), ()),
     'StataLexer': ('pygments.lexers.stata', 'Stata', ('stata', 'do'), ('*.do', '*.ado'), ('text/x-stata', 'text/stata', 'application/x-stata')),
     'SuperColliderLexer': ('pygments.lexers.supercollider', 'SuperCollider', ('supercollider', 'sc'), ('*.sc', '*.scd'), ('application/supercollider', 'text/supercollider')),
+    'SwanLexer': ('pygments.lexers.swan', 'Swan', ('swan',), ('*.swan', '*.swani'), ()),
```
*.swant should be added

ansbmartin replied:
done

xfornari commented on pygments/lexers/swan.py:
```diff
@@ -0,0 +1,118 @@
+            (words((
+                # declarations
+                'function', 'node', 'sensor', 'type', 'const',
+                'guarantee', 'returns', 'specialize', 'where',
```
guarantee appears twice, missing assert

ansfcouadau replied:
Good catch (although I see `assert` in the list).
Should we locally run the `tox` tests with a full module from the CruiseControl demo model to make sure we have no missing keyword?

ansbmartin replied:
Fixed duplicate `guarantee`. I also checked the keyword list against the parser

xfornari commented on pygments/lexers/swan.py:
```diff
@@ -0,0 +1,118 @@
+                'if', 'then', 'else', 'case', 'of',
+                'last', 'pre',
+                'reverse', 'pack', 'transpose', 'window', 'flatten',
+                'merge', 'mod', 'self', 'when', 'with', 'default',
```
when should be with match ?

ansbmartin replied:
done

xfornari commented on pygments/lexers/swan.py:
```diff
@@ -0,0 +1,118 @@
+            (r'#[\w\-]+', token.Name.Label),  # LUID
+            (r'[a-zA-Z]\w*', token.Name), # ID
+            (r'\s', token.Text),
+            (r'\S', token.Text)
```
{*% .. %*}  is missing (quite a lot when reading a .swan from Editor (errors, or special markups)

ansbmartin replied:
added as `token.Comment.Special`

ansfcouadau commented on pygments/lexers/swan.py:
```diff
@@ -0,0 +1,118 @@
+
+class SwanLexer(RegexLexer):
+    """
+    Lexer for the Swan programming language (version 2025).
```
The linked language primer document mentions two versions: "2025.0" (for the language I imagine) and "2.1" for the document itself. Perhaps we should pick one and use it in this comment (e.g. "version 2025.0")?

ansbmartin replied:
I removed the version from the comment, as the lexer is meant to support any version of the Swan language

ansfcouadau commented on pygments/lexers/swan.py:
```diff
@@ -0,0 +1,118 @@
+    url = 'https://github.com/ansys/DevRelPublic/blob/main/Downloads/ScadeOne/SwanPrimer_Rev_2_1.pdf'
+    aliases = ['swan', 'scade one']
+    filenames = ['*.swan', '*.swani']
+    version_added = '2.20'
```
Same comment on version, we should stick to the same version number everywhere to be consistent.

ansbmartin replied:
same as above

ansfcouadau reviewed (changes requested):
Aside from version number consistency, it looks good to me!

jfthuong commented:
How about putting an example in "tests/examplefiles"?

jfthuong reviewed (commented):
How about putting an example in "_tests/examplefiles_"?

jfthuong commented on pygments/lexers/swan.py:
```diff
@@ -112,6 +112,7 @@ class SwanLexer(RegexLexer):
             (r'/\*', token.Comment.Multiline, 'comment'),  # must be before operators
             (r'<<', token.Generic.Strong, 'chevron'),      # must be before operators
             (r'--.*$', token.Comment.Singleline),          # must be before operators
+            (r"{[a-zA-Z0-9_]*%", token.Comment.Special, 'markup'),
```
Maybe we could escape the `{` for a slightly cleaner regex `r"\{[a-zA-Z0-9_]*%"`

ansbmartin replied:
done

ansbmartin commented:
Thanks for the reviews and approvals, I have created the PR on the upstream repository: https://github.com/pygments/pygments/pull/2878

thavelick committed:
fix lua regex causing runaway backtracking.

Use negative lookahead to avoid runway backtracking in whitespace.

Fixes #2839

Commit:
addex *.xbrl mappings

d-torrance committed:
Update symbols for Macaulay2 1.25.05

rrthomas committed:
autopygmentize: support more file types

Improves support for XML and Python, adds support for CSS, Go and Rust.

rrthomas committed:
autopygmentize: make the fallthrough to use pygmentize -N universal

Rather than just trying it for files diagnosed as text/plain by the main
command (file(1) by default), try it for any otherwise unrecognized MIME
type.

This fixes support for e.g. Markdown.

rrthomas committed:
rrt style: add more styles for Generic and Name.Tag

Primarily aimed at Markdown support.

omnivelociraptor committed:
Add rules for interpolated strings in MoonScript lexer

Anteru committed:
Merge pull request #2882 from thavelick/fix_lua_runaway_regex

fix lua regex causing runaway backtracking.

Anteru committed:
Update CHANGES.

technomancy committed:
Update lexer for Fennel.

* Add new filename pattern for macro modules.
* Add a mime type (Trac ignores languages that lack this for some reason).
* Update list of special forms for the latest Fennel version.
* Add backtick as a punctuation character.

technomancy committed:
Update mapping file for new Fennel extension and mime type.

Anteru committed:
Update ruff version.

Need to disable the format string rule as we're failing that, a lot.

Anteru committed:
Disable pyodide (currently broken.)

Anteru committed:
Prepare v2.19.2 release.

SpecLad committed:
Switch license metadata to the PEP 639 format

License classifiers and the table form of `license` are now deprecated.

dnicolodi committed:
meson: Drop spurious ] in operators definition

dnicolodi committed:
meson: Add *, /, % operators definition

Fixes #2918.

DanielEScherzer committed:
PHP lexer: add `__PROPERTY__` magic constant, update link for source

Spill-Tea committed:
feat(lexers.python): Include additional keywords, and builtins. Separate Keyword.Constant vs Name.Builtin.Pseudo tokens.

Spill-Tea committed:
feat(lexers.python): Update cdef context of cython lexer to differentiate function, class, and variable name tokens. #2932

Spill-Tea committed:
feat(lexers.cython): Include cppclass keyword in context where cdef is not declared first. #2932

Spill-Tea committed:
feat(snippts): Add cython snippets to repository. #2932

Spill-Tea committed:
feat(lexers.python): Add missing builtin type double. Add missing keywords 'namespace' and 'noexcept' to cython lexer. #2932

Spill-Tea committed:
feat(snippets.cython): Include additional testing snippets to improve coverage of builtins, keywords, and docstrings.

Spill-Tea committed:
feat(lexers.python): Remove NotImplemented from named exceptions, due to unreachable redundancy in builtin pseudo token.

Spill-Tea committed:
fix(snippets.cython): Update cppclass test snippet due to updated namespace keyword.

Anteru committed:
Properly disable the demo link.

Anteru committed:
Merge pull request #2894 from rrthomas/autopygmentize-update

autopygmentize update

Anteru committed:
Merge pull request #2933 from Spill-Tea/cython_lexer_update

Cython lexer update

Anteru committed:
Merge pull request #2916 from SpecLad/pep-639

Switch license metadata to the PEP 639 format

Anteru committed:
Merge pull request #2911 from technomancy/update-fennel

Update lexer for Fennel.

Anteru committed:
Merge pull request #2907 from omnivelociraptor/cm-moonscript-lexer-string-interpol

Add rules for interpolated strings in MoonScript lexer

Anteru committed:
Merge pull request #2893 from d-torrance/macaulay2

Update symbols for Macaulay2 1.25.05

Anteru committed:
Merge pull request #2895 from rrthomas/rrt-style-update

rrt style: add more styles for Generic and Name.Tag

Anteru committed:
Merge pull request #2919 from dnicolodi/meson-operators

meson: Add *, /, % operators definition

Anteru committed:
Merge pull request #2924 from DanielEScherzer/patch-1

PHP lexer: add `__PROPERTY__` magic constant, update link for source

Anteru committed:
Merge pull request #2891 from lor1113/xbrl

added *.xbrl mappings
